docs: improve typings docs for custom matchers (#12195)
facebook/jest: docs/ExpectAPI.md

});
```

_Note_: In TypeScript, when using `@types/jest` for example, you can declare the new `toBeWithinRange` matcher in the imported module like this (`toBeWithinRange.ts`):

> ```ts

Reviewer: ```suggestion
```ts title="toBeWithinRange.ts"
```
Perhaps?

expect.extend({

>   toBeWithinRange(received, floor, ceiling) {

Reviewer: ```suggestion
  toBeWithinRange(received: number, floor: number, ceiling: number) {
```
I like the idea to include this snippet here, should it be typed too?

}
```

> Depending on your TypeScript setup, you may also need this:

Reviewer: It might be useful to add more details on the specific setup which requires this line. At the moment it looks puzzling. If `jest` is exported here, do you import it later to another module? Or perhaps simply an export is needed, because of `"isolatedModules": true` setting?

Author: > At the moment it looks puzzling

Absolutely.

Currently, you need `export = jest` when you create a new react app with the typescript template and the default configuration. IMO it's already an improvement to help people find a solution to this problem, even if it is "depending on your TypeScript setup".

`npx create-react-app my-app --template typescript`

```json
{
  "compilerOptions": {
    "target": "es5",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "allowJs": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noFallthroughCasesInSwitch": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "jsx": "react-jsx"
  },
  "include": [
    "src"
  ]
}
```

> It might be useful to add more details on the specific setup which requires this line. At the moment it looks puzzling. If jest is exported here, do you import it later to another module? Or perhaps simply an export is needed, because of "isolatedModules": true setting?

If you know _exactly_ when this is needed (and when not), it could be documented here. But I don't know about all the configurations which do / don't need this, so I don't feel confident in documenting it like that. Also, I don't know when this stopped working, assuming that it worked before...

Reviewer: Hm.. Do you *import* the file with the declaration as current docs suggest: "declare the new matcher in the *imported* module". Or do you include the declaration in a `.d.ts` file? The thing is that declaration merging does not work in TypeScript without importing and `.d.ts` is not valid without `export` (an empty `export {}` does the job). Just guessing around.

Reviewer: > Hm.. Do you _import_ the file with the declaration as current docs suggest: "declare the new matcher in the _imported_ module". Or do you include the declaration in a `.d.ts` file? The thing is that declaration merging does not work in TypeScript without importing and `.d.ts` is not valid without `export` (an empty `export {}` does the job) (it might be I am wrong here). Just guessing around.

Author: > it might be I am wrong here

No, you are right!

I have the type declarations in a separate `.d.ts` file, and your suggestion works:

`types/jest/index.d.ts`

```ts
declare global {
  namespace jest {
    interface Matchers<R> {
      toDoSomething(): R
    }
  }
}

export {}
```

I'll update the PR accordingly. Thanks! 🏅

Author: @mrazauskas I updated the commit, feel free to review it :)

Reviewer: Done (;

```ts

> export = jest

Reviewer 2: ```suggestion
export = jest;
```

}
```

> If you want to move the typings to a separate file (e.g. `types/jest/index.d.ts`), you may need to an export, e.g.:
> 
> ```ts
> declare global {
>   namespace jest {
>     interface Matchers<R> {
>       toBeWithinRange(a: number, b: number): R;
>     }
>   }
> }
> 
> export {};
> ```

Reviewer: ```suggestion
If you prefer to keep the typings in a `.d.ts` declaration file, you may need to include an empty `export {}`.
```

Just thinking out loud. Seems like there is no need to repeat the code again. Or my version is less clear? Feel free to rework it.
